server/quoridor/quoridor.go

```diff
@@ -0,0 +1,444 @@
+	game.CurrentTurn = PlayerPosition(next)
+}
+
+func (game *Game) IsGameOver() bool {
+	return !game.EndDate.IsZero()
+}
+
+func (game *Game) MaybeReturnWinnerPlayerPosition() PlayerPosition {
+	for position, player := range game.Players {
+		winningPosition := winningPositions[position]
+		if player.Pawn.Position.Y == winningPosition.Y ||
+			player.Pawn.Position.X == winningPosition.X {
+			return position
+		}
+	}
+	return -1
+}
+
+// Deep copy of a game. Will copy all Players, Pieces, and Board.
+func (game *Game) Copy() Game {
+	newGame := Game{
+		Id:          game.Id,
+		CurrentTurn: game.CurrentTurn,
+		StartDate:   game.StartDate,
+		EndDate:     game.EndDate,
+		Winner:      game.Winner,
+		Name:        game.Name,
+	}
+
+	newGame.Board = make(Board)
+	for pos, piece := range game.Board {
+		newPiece := piece
+		newGame.Board[pos] = newPiece
+	}
+
+	newGame.Players = make(map[PlayerPosition]*Player)
+	for pos, player := range game.Players {
+		newGame.Players[pos] = &Player{
+			Barriers:   player.Barriers,
+			Pawn:       player.Pawn,
+			PlayerId:   player.PlayerId,
+			PlayerName: player.PlayerName,
+		}
+	}
+	return newGame
+}
+
+// Starts a game by setting the StartDate to the current instant of time. Returns an error if there aren't enough
+// players, or the game has already started.
+func (game *Game) StartGame() error {
+	if !(len(game.Players) == 2 || len(game.Players) == 4) {
+		return errors.New(fmt.Sprintf("can't start game, wrong number of players (%d)", len(game.Players)))
+	}
+	if !game.StartDate.IsZero() {
+		return errors.New(fmt.Sprintf("game already started"))
+	}
+	game.StartDate = time.Now()
+	return nil
+}
+
+// Initialize with default values
+func NewGame(id uuid.UUID, name string) (*Game, error) {
+	if id == uuid.Nil {
+		return nil, errors.New("unable to create game, need valid id")
+	}
+	if name == "" {
+		return nil, errors.New("unable to create game, need non-empty name")
+	}
+	return &Game{
+		Board:       make(map[Position]Piece),
+		Players:     make(map[PlayerPosition]*Player),
+		CurrentTurn: PlayerOne,
+		Id:          id,
+		Name:        name,
+		Winner:      -1,
+	}, nil
+}
+
+// Return the state of the Board after the move, of the current board if an error
+func (game *Game) MovePawn(newPosition Position, player PlayerPosition) error {
+	pawn := &game.Players[player].Pawn
+	if !IsValidPawnLocation(newPosition) {
+		return errors.New("invalid Pawn Location")
+	}
+	if game.CurrentTurn != player {
+		return errors.New(fmt.Sprintf("wrong turn, current turn is for Player: %d", game.CurrentTurn))
+	}
+	if moveError := isValidPawnMove(newPosition, pawn.Position, game.Board); moveError != nil {
+		return moveError
+	}
+	delete(game.Board, pawn.Position)
+	pawn.Position = newPosition
+	game.Board[pawn.Position] = *pawn
+	checkGameOver(game)
+	game.NextTurn()
+	return nil
+}
+
+func IsValidPawnLocation(position Position) bool {
+	return position.X%2 == 0 && position.Y%2 == 0
```

Comment:
bounds checking too?

Reply:
Bounds checking is done in isOnBoard, and not unique to a Pawn location. The idea here is that the error message generated for MovePawn to an odd position results in "invalid pawn location" rather than "the pawn cannot reach that square." 

Maybe I should surface the IsOnBoard check to an explicit, top-level error.